File: 2Lab/labs5/lab5.cpp
#include <iostream>
#include <cmath>
#include <vector>
using namespace std;

const double PI = 3.14159265358979323846;


class Figure {
protected:
    double length;
    double width;

public:
    Figure() : length(0), width(0) {}
    Figure(double l, double w) : length(l), width(w) {}

    double getLength() const { return length; }
    double getWidth() const { return width; }

    virtual double area() const {
        return length * width;
    }

    virtual double perimeter() const {
        return 2 * (length + width);
    }

    virtual ~Figure() {}
};

class Ellipse : public Figure {
public:
    Ellipse() : Figure() {}
    Ellipse(double a, double b) : Figure(a, b) {}

    double area() const override {
        return PI * length * width;
    }

    double perimeter() const override {
        return PI * (3 * (length + width) - sqrt((3 * length + width) * (length + 3 * width)));
    }
};

class Circle : public Figure {
public:
    Circle() : Figure() {}
    Circle(double r) : Figure(r, r) {}

    double area() const override {
        return PI * length * length;
    }

    double perimeter() const override {
        return 2 * PI * length;
    }
};

void displayFigureInfo(const Figure& fig) {
    cout << "Area: " << fig.area() << endl;
    cout << "Perimeter: " << fig.perimeter() << endl;
}

int main() {
    Ellipse e(3.0, 2.0);
    Circle c(5.0);

    cout << "--- Ellipse ---" << endl;
    displayFigureInfo(e);

    cout << "--- Circle ---" << endl;
    displayFigureInfo(c);

    return 0;
}

File: 2Lab/labs5/lab5.cs
using System;

namespace GeometryPolymorphism
{
    public class Figure
    {
        protected double Length { get; }
        protected double Width { get; }

        public Figure(double length, double width)
        {
            Length = length;
            Width = width;
        }

        public virtual double Area()
        {
            return Length * Width;
        }

        public virtual double Perimeter()
        {
            return 2 * (Length + Width);
        }
    }

    public class Ellipse : Figure
    {
        public Ellipse(double a, double b) : base(a, b) { }

        public override double Area()
        {
            return Math.PI * Length * Width;
        }

        public override double Perimeter()
        {
            return Math.PI * (3 * (Length + Width) - Math.Sqrt((3 * Length + Width) * (Length + 3 * Width)));
        }
    }

    public class Circle : Figure
    {
        public Circle(double radius) : base(radius, radius) { }

        public override double Area()
        {
            return Math.PI * Length * Length;
        }

        public override double Perimeter()
        {
            return 2 * Math.PI * Length;
        }
    }

    class Program
    {
        static void DisplayFigureInfo(Figure figure)
        {
            Console.WriteLine($"Area: {figure.Area()}");
            Console.WriteLine($"Perimeter: {figure.Perimeter()}");
        }

        static void Main(string[] args)
        {
            Ellipse e = new Ellipse(3.0, 2.0);
            Circle c = new Circle(5.0);

            Console.WriteLine("--- Ellipse ---");
            DisplayFigureInfo(e);

            Console.WriteLine("--- Circle ---");
            DisplayFigureInfo(c);
        }
    }
}
